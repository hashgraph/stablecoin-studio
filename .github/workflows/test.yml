name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test