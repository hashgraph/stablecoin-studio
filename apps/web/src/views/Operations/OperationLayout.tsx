import type { ReactNode } from 'react';
import type { ButtonProps as ChakraButtonProps } from '@chakra-ui/react';
import { Button, Flex, Stack, Heading, SimpleGrid } from '@chakra-ui/react';
import { useNavigate } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { RouterManager } from '../../Router/RouterManager';
import BaseContainer from '../../components/BaseContainer';
import DetailsReview from '../../components/DetailsReview';
import { SELECTED_WALLET_COIN } from '../../store/slices/walletSlice';
import { useRefreshCoinInfo } from '../../hooks/useRefreshCoinInfo';

export interface OperationLayoutProps {
	LeftContent: ReactNode;
	onConfirm: () => void;
	confirmBtnProps?: ChakraButtonProps;
}

const OperationLayout = ({ LeftContent, onConfirm, confirmBtnProps }: OperationLayoutProps) => {
	const navigate = useNavigate();
	const { t } = useTranslation(['operations', 'global']);
	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);
	const unknown = t('global:common.unknown');

	const handleGoBack = () => {
		RouterManager.goBack(navigate);
	};

	useRefreshCoinInfo();
	const optionalDetailsFinite = [
		{
			label: t('operations:details.initialSupply'),
			value: selectedStableCoin?.initialSupply ? selectedStableCoin?.initialSupply : unknown,
		},
		{
			label: t('operations:details.totalSupply'),
			value: selectedStableCoin?.totalSupply ? selectedStableCoin?.totalSupply : unknown,
		},
		{
			label: t('operations:details.maxSupply'),
			value: selectedStableCoin?.maxSupply ? selectedStableCoin?.maxSupply : unknown,
		},
		{
			label: t('operations:details.supplyType'),
			value: selectedStableCoin?.maxSupply?.isZero()
				? t('operations:details.infinite')
				: t('operations:details.finite'),
		},
	];

	const optionalDetailsInfinite = [
		{
			label: t('operations:details.initialSupply'),
			value: selectedStableCoin?.initialSupply ? selectedStableCoin.initialSupply : unknown,
		},
		{
			label: t('operations:details.totalSupply'),
			value: selectedStableCoin?.totalSupply ? selectedStableCoin?.totalSupply : unknown,
		},
		{
			label: t('operations:details.supplyType'),
			value: selectedStableCoin?.maxSupply?.isZero()
				? t('operations:details.infinite')
				: t('operations:details.finite'),
		},
		{
			label: t('operations:details.metadata'),
			value: selectedStableCoin?.metadata,
		},
	];

	return (
		<BaseContainer title={t('global:operations.title')}>
			<Flex
				direction='column'
				bg='brand.gray100'
				px={{ base: 4, lg: 14 }}
				pt={{ base: 4, lg: 14 }}
				pb={6}
			>
				<SimpleGrid columns={{ lg: 2 }} gap={{ base: 4, lg: 20 }}>
					<Stack>{LeftContent}</Stack>
					<Stack bg='brand.white' p={6}>
						<Stack bg='brand.white' spacing={10}>
							<Heading fontSize='16px' color='brand.secondary' data-testid='details-title'>
								{t('operations:details.title')}
							</Heading>

							<DetailsReview
								title={t('operations:details.basicTitle')}
								titleProps={{ fontWeight: 700, color: 'brand.secondary' }}
								details={[
									{
										label: t('operations:details.name'),
										value: selectedStableCoin?.name || unknown,
									},
									{
										label: t('operations:details.symbol'),
										value: selectedStableCoin?.symbol || unknown,
									},
									{
										label: t('operations:details.decimals'),
										value: selectedStableCoin?.decimals || unknown,
									},
								]}
							/>
							<DetailsReview
								title={t('operations:details.optionalTitle')}
								titleProps={{ fontWeight: 700, color: 'brand.secondary' }}
								details={
									selectedStableCoin?.maxSupply?.isZero()
										? optionalDetailsInfinite
										: optionalDetailsFinite
								}
							/>
						</Stack>
					</Stack>
				</SimpleGrid>
				<Flex justify='flex-end' pt={6}>
					<Stack direction='row' spacing={6}>
						<Button data-testid='cancel-btn' onClick={handleGoBack} variant='secondary'>
							{t('global:common.goBack')}
						</Button>
						<Button data-testid='confirm-btn' onClick={onConfirm} {...confirmBtnProps}>
							{t('global:common.accept')}
						</Button>
					</Stack>
				</Flex>
			</Flex>
		</BaseContainer>
	);
};

export default OperationLayout;
