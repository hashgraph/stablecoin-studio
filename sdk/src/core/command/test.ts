/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
