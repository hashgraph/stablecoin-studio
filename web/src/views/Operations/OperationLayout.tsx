import type { ReactNode } from 'react';
import type { ButtonProps as ChakraButtonProps } from '@chakra-ui/react';
import { Button, Flex, Stack, Heading, SimpleGrid } from '@chakra-ui/react';
import { useNavigate } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { RouterManager } from '../../Router/RouterManager';
import BaseContainer from '../../components/BaseContainer';
import DetailsReview from '../../components/DetailsReview';
import { SELECTED_WALLET_COIN } from '../../store/slices/walletSlice';
import { formatAmountWithDecimals } from '../../utils/inputHelper';

export interface OperationLayoutProps {
	LeftContent: ReactNode;
	onConfirm: () => void;
	confirmBtnProps?: ChakraButtonProps;
}

const OperationLayout = ({ LeftContent, onConfirm, confirmBtnProps }: OperationLayoutProps) => {
	const navigate = useNavigate();
	const { t } = useTranslation(['operations', 'global']);
	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);
	const unknown = t('global:common.unknown');

	const handleGoBack = () => {
		RouterManager.goBack(navigate);
	};

	return (
		<BaseContainer title={t('global:operations.title')}>
			<Flex
				direction='column'
				bg='brand.gray100'
				px={{ base: 4, lg: 14 }}
				pt={{ base: 4, lg: 14 }}
				pb={6}
			>
				<SimpleGrid columns={{ lg: 2 }} gap={{ base: 4, lg: 20 }}>
					<Stack>{LeftContent}</Stack>
					<Stack bg='brand.white' p={6}>
						<Stack bg='brand.white' spacing={10}>
							<Heading fontSize='16px' color='brand.secondary' data-testid='details-title'>
								{t('operations:details.title')}
							</Heading>

							<DetailsReview
								title={t('operations:details.basicTitle')}
								titleProps={{ fontWeight: 700, color: 'brand.secondary' }}
								details={[
									{
										label: t('operations:details.name'),
										value: selectedStableCoin?.name || unknown,
									},
									{
										label: t('operations:details.symbol'),
										value: selectedStableCoin?.symbol || unknown,
									},
									{
										label: t('operations:details.decimals'),
										value: selectedStableCoin?.decimals || unknown,
									},
								]}
							/>
							<DetailsReview
								title={t('operations:details.optionalTitle')}
								titleProps={{ fontWeight: 700, color: 'brand.secondary' }}
								details={[
									{
										label: t('operations:details.initialSupply'),
										value: selectedStableCoin?.initialSupply
											? formatAmountWithDecimals({
													amount: Number(selectedStableCoin?.initialSupply),
													decimals: selectedStableCoin?.decimals || 0,
											  })
											: unknown,
									},
									{
										label: t('operations:details.totalSupply'),
										value: selectedStableCoin?.totalSupply
											? formatAmountWithDecimals({
													amount: Number(selectedStableCoin?.totalSupply),
													decimals: selectedStableCoin?.decimals || 0,
											  })
											: unknown,
									},
									{
										label: t('operations:details.supplyType'),
										// @ts-ignore Property 'supplyType' does not exist on type 'IStableCoinDetail'.
										value:
											selectedStableCoin?.maxSupply === ('0' as unknown as BigInt)
												? t('operations:details.infinite')
												: t('operations:details.finite'),
									},
								]}
							/>
						</Stack>
					</Stack>
				</SimpleGrid>
				<Flex justify='flex-end' pt={6}>
					<Stack direction='row' spacing={6}>
						<Button data-testid='cancel-btn' onClick={handleGoBack} variant='secondary'>
							{t('global:common.goBack')}
						</Button>
						<Button data-testid='confirm-btn' onClick={onConfirm} {...confirmBtnProps}>
							{t('global:common.accept')}
						</Button>
					</Stack>
				</Flex>
			</Flex>
		</BaseContainer>
	);
};

export default OperationLayout;
