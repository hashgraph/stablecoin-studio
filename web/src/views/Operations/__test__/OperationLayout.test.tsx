import userEvent from '@testing-library/user-event';
import { render } from '../../../test/index';
import OperationLayout, { OperationLayoutProps } from '../OperationLayout';

describe(`<${OperationLayout.name} />`, () => {
	const props: OperationLayoutProps = {
		LeftContent: <span data-testid='left-content'></span>,
		RightContent: <span data-testid='right-content'></span>,
		onConfirm: jest.fn(),
	};

	test('should render correctly', () => {
		const component = render(<OperationLayout {...props} />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should render leftContent & rightContent', () => {
		const component = render(<OperationLayout {...props} />);

		expect(component.getByTestId('left-content')).toBeInTheDocument();
		expect(component.getByTestId('right-content')).toBeInTheDocument();
	});

	test('should call onConfirm when confirm btn is clicked', () => {
		const component = render(<OperationLayout {...props} />);

		userEvent.click(component.getByTestId('confirm-btn'));

		expect(props.onConfirm).toHaveBeenCalled();
	});

	test.todo('should go to operations route when cancel btn is clicked');
});
