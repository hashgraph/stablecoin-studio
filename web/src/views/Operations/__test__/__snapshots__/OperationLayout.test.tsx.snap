// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OperationLayout /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-16oka5y"
  >
    <h2
      class="chakra-heading css-llrpa2"
      data-testid="base-container-heading"
    >
      Operations
    </h2>
    <div
      class="css-1hkemoa"
      data-testid="base-container-children"
    >
      <div
        class="css-1us6gcn"
      >
        <div
          class="css-1r9sqqd"
        >
          <div
            class="chakra-stack css-n21gh5"
          >
            <span
              data-testid="left-content"
            />
          </div>
          <div
            class="chakra-stack css-eunu1i"
          >
            <div
              class="chakra-stack css-wqnj46"
            >
              <h2
                class="chakra-heading css-12c29ps"
                data-testid="details-title"
              >
                Stable coin details
              </h2>
              <div
                class="css-13brihr"
              >
                <p
                  class="chakra-text css-53sllx"
                  data-testid="details-review-title"
                >
                  Basic details
                </p>
                <div
                  class="css-0"
                >
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-gg4vpm"
                      data-testid="details-review-detail-0"
                    >
                      <p
                        class="chakra-text css-rhg419"
                      >
                        Name
                      </p>
                      <p
                        class="chakra-text css-rhg419"
                      >
                        Example
                      </p>
                    </div>
                    <hr
                      aria-orientation="horizontal"
                      class="chakra-divider css-zaaihy"
                      title="divider"
                    />
                  </div>
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-gg4vpm"
                      data-testid="details-review-detail-1"
                    >
                      <p
                        class="chakra-text css-rhg419"
                      >
                        Symbol
                      </p>
                      <p
                        class="chakra-text css-rhg419"
                      >
                        example
                      </p>
                    </div>
                    <hr
                      aria-orientation="horizontal"
                      class="chakra-divider css-zaaihy"
                      title="divider"
                    />
                  </div>
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-gg4vpm"
                      data-testid="details-review-detail-2"
                    >
                      <p
                        class="chakra-text css-rhg419"
                      >
                        Decimals
                      </p>
                      <p
                        class="chakra-text css-rhg419"
                      >
                        6
                      </p>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="css-13brihr"
              >
                <p
                  class="chakra-text css-53sllx"
                  data-testid="details-review-title"
                >
                  Optional details
                </p>
                <div
                  class="css-0"
                >
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-gg4vpm"
                      data-testid="details-review-detail-0"
                    >
                      <p
                        class="chakra-text css-rhg419"
                      >
                        Initial supply
                      </p>
                      <p
                        class="chakra-text css-rhg419"
                      >
                        20
                      </p>
                    </div>
                    <hr
                      aria-orientation="horizontal"
                      class="chakra-divider css-zaaihy"
                      title="divider"
                    />
                  </div>
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-gg4vpm"
                      data-testid="details-review-detail-1"
                    >
                      <p
                        class="chakra-text css-rhg419"
                      >
                        Total supply
                      </p>
                      <p
                        class="chakra-text css-rhg419"
                      >
                        200
                      </p>
                    </div>
                    <hr
                      aria-orientation="horizontal"
                      class="chakra-divider css-zaaihy"
                      title="divider"
                    />
                  </div>
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-gg4vpm"
                      data-testid="details-review-detail-2"
                    >
                      <p
                        class="chakra-text css-rhg419"
                      >
                        Supply type (infinite or finite)
                      </p>
                      <p
                        class="chakra-text css-rhg419"
                      >
                        infinite
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-3gu2fl"
        >
          <div
            class="chakra-stack css-1amf41o"
          >
            <button
              class="chakra-button css-6blmsx"
              data-testid="cancel-btn"
              type="button"
            >
              Go back
            </button>
            <button
              class="chakra-button css-16vubea"
              data-testid="confirm-btn"
              type="button"
            >
              Accept
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
